### ISSUE TITLE
Increase the usage of augmented assignment statements

### ISSUE THREAD

**Issue by elfring** [librosa/librosa#1403]
:eyes: Some source code analysis tools can help to find opportunities for improving software components.
:thought_balloon: I propose to [increase the usage of augmented assignment statements](https://docs.python.org/3/reference/simple_stmts.html#augmented-assignment-statements "Augmented assignment statements") accordingly.

Would you like to integrate anything from a transformation result which can be generated by a command like the following?
(:point_right: Please check also for questionable change suggestions because of an evolving search pattern.)
```
lokal$ perl -p -i.orig -0777 -e 's/^(?<indentation>\s+)(?<target>\S+)\s*=\s*\k<target>[ \t]*(?<operator>[+\-%&|^@]|\*\*?|\/\/?|<<|>>)/$+{indentation}$+{target} $+{operator}=/gm' $(find ~/Projekte/librosa/lokal -name '*.py')
```

**Comment by bmcfee** [comment]
What benefit do you see in doing this?

I tend to avoid using these for two reasons:

- it's often clearer to read "x = x + y" than "x += y"
- more importantly, the semantics are not generally identical: one can modify a variable in place, the other creates a new variable. Often in numerical python code, the latter should be preferred so that functions do not unexpectedly modify their inputs. (Remember, we don't have the equivalent of `const` or `val` here.). Sometimes an in place modification is okay, but I wouldn't trust a regular expression to distinguish these cases.

**Comment by elfring** [comment]
>more importantly, the semantics are not generally identical: …

* I would expect that the logical semantics must be identical in the sense that something gets assigned to a selected variable.
* The software run time characteristics can be different.

Do you find [the explanation by the Python enhancement proposal 203](https://www.python.org/dev/peps/pep-0203/#rationale "Proposal for augmented assignments") (from 2000-07-13) insufficient? :thinking:


>Sometimes an in place modification is okay,

:thought_balloon: Can such applications be extended?


>but I wouldn't trust a regular expression to distinguish these cases.

This view is reasonable.
Would you eventually like to adjust the distinction where in-place modifications should be excluded?

**Comment by bmcfee** [comment]
> I would expect that the logical semantics must be identical in the sense that something gets assigned to a selected variable.

Except that it isn't the same.  Consider the following two functions:

```python
def f(x):
    # x is assumed to be a numpy.ndarray
    x = x + 1
    return x

def g(x):
    x += 1
    return x
```
And now call them as follows:
```python
In [4]: x = np.arange(3)

In [5]: x
Out[5]: array([0, 1, 2])

In [6]: f(x)
Out[6]: array([1, 2, 3])

In [7]: x
Out[7]: array([0, 1, 2])

In [8]: g(x)
Out[8]: array([1, 2, 3])

In [9]: x
Out[9]: array([1, 2, 3])
```
The first function leaves the input unchanged, the second does not.  (This is because ndarrays implement the `+=` operator distinctly from the `+` operator.)  In functional programming speak, we would say that the second has side-effects and the first is a pure function.  We generally try to avoid side effects.


> The software run time characteristics can be different.

Sure, but correctness is always more important than speed.


> Do you find [the explanation by the Python enhancement proposal 203](https://www.python.org/dev/peps/pep-0203/#rationale) (from 2000-07-13) insufficient? thinking

Yes, actually.  The question is not about the syntactic difference, but rather what is to be gained in our specific context.

**Comment by elfring** [comment]
>Sure, but correctness is always more important …

Some users expect that the applied software would usually be executed in the correct way.
They occasionally take also another look at “side effects”.
* Execution durations
* Memory allocations


>The question is not about the syntactic difference, but rather what is to be gained in our specific context.

Can any differences become more relevant for the observed software run time characteristics? :thinking:

**Comment by bmcfee** [comment]
Sorry, I think we're possibly speaking past each other here.

You're proposing a fairly general and abstract change, and my response is that the proposed change, in general, may not be a good thing.  I definitely agree that there could be specific instances where this change might be appropriate, and we could almost certainly gain some efficiency in places.

My point is that this needs to be handled on a case-by-case basis, referring to specific parts of the code, not as a general search-and-replace.

**Comment by elfring** [comment]
>My point is that this needs to be handled on a case-by-case basis, referring to specific parts of the code,

:thought_balloon: Would you get into the mood to take another look at data processing results from an adjusted Perl command example?


>not as a general search-and-replace.

Would you become interested in more sophisticated source code analyses and transformation approaches? :thinking:

**Comment by bmcfee** [comment]
> thought_balloon Would you get into the mood to take another look at data processing results from an adjusted Perl command example?

I'm happy to look at specific cases, but I would prefer to not sift through every example of a binary operation.



> Would you become interested in more sophisticated source code analyses and transformation approaches? thinking

If it's in service of a specific improvement, sure.  But I hope you appreciate that I'm wary of throwing source code into abstract tooling.  Our code is not perfect, but it is generally pretty well tested and thought out, and we do use basic analyses (linting, etc) fairly often.

**Comment by elfring** [comment]
>I'm happy to look at specific cases, …

The shown Perl command example can point places out for further development considerations.
:thought_balloon: 28 source code transformations might look reasonable, don't they?

**Comment by bmcfee** [comment]
> The shown Perl command example can point places out for further development considerations.

I understand that you're trying to be helpful here, but here's how it looks from the maintainer's perspective:

- Your initial post proposes a change to coding style, but does not refer to any specific problems that would be addressed by this proposal.
- The proposal does not identify specific parts of the code that would need changing.
- You're asking maintainers to run a fairly inscrutable perl script to *possibly* identify parts of the code that *might* be candidates for rewriting.

In short, you're asking maintainers to do all the work here, and not providing any compelling reason to do so.  This is not a great way to make friends in any open source community.

As I said before, I'll entertain specific suggestions with good justification, but I see no reason to pursue this proposal on my own time.

**Comment by elfring** [comment]
>Your initial post proposes a change to coding style,

I dared to point another adjustment possibility out.


>but does not refer to any specific problems that would be addressed by this proposal.

This view might be influenced by different interpretations according to the term “augmented assignment statement”.


>The proposal does not identify specific parts of the code that would need changing.

Indirectly. ‒ I offered a corresponding Perl command example for this purpose.


>You're asking maintainers to run a fairly inscrutable perl script to _possibly_ identify parts of the code that _might_ be candidates for rewriting.

Interested developers can try the shown search and replace approach out, can't they?

Would you find the following data format representation more appealing?

```diff
diff --git a/librosa/beat.py b/librosa/beat.py
index 3536d9074..8f715c8d6 100644
--- a/librosa/beat.py
+++ b/librosa/beat.py
@@ -634,7 +634,7 @@ def __normalize_onsets(onsets):
 
     norm = onsets.std(ddof=1)
     if norm > 0:
-        onsets = onsets / norm
+        onsets /= norm
     return onsets
 
 
@@ -669,7 +669,7 @@ def __beat_track_dp(localscore, period, tightness):
 
         # Search over all possible predecessors
         candidates = txwt.copy()
-        candidates[z_pad:] = candidates[z_pad:] + cumscore[window[z_pad:]]
+        candidates[z_pad:] += cumscore[window[z_pad:]]
 
         # Find the best preceding beat
         beat_location = np.argmax(candidates)
@@ -685,7 +685,7 @@ def __beat_track_dp(localscore, period, tightness):
             first_beat = False
 
         # Update the time range
-        window = window + 1
+        window += 1
 
     return backlink, cumscore
 
diff --git a/librosa/core/audio.py b/librosa/core/audio.py
index bd0e6f160..74c22069c 100644
--- a/librosa/core/audio.py
+++ b/librosa/core/audio.py
@@ -203,7 +203,7 @@ def __audioread_load(path, offset, duration, dtype):
         for frame in input_file:
             frame = util.buf_to_float(frame, dtype=dtype)
             n_prev = n
-            n = n + len(frame)
+            n += len(frame)
 
             if n < s_start:
                 # offset is after the current frame
@@ -710,7 +710,7 @@ def get_duration(
 
         # If centered, we lose half a window from each end of S
         if center:
-            n_samples = n_samples - 2 * int(n_fft // 2)
+            n_samples -= 2 * int(n_fft // 2)
 
     else:
         n_samples = y.shape[-1]
@@ -975,8 +975,8 @@ def __lpc(y, order, ar_coeffs, ar_coeffs_prev, reflect_coeff, den, epsilon):
         # eqns 13 and 14.  We start with f_{M-1,k+1} and b_{M-1,k} and use them
         # to compute f_{M,k} and b_{M,k}
         fwd_pred_error_tmp = fwd_pred_error
-        fwd_pred_error = fwd_pred_error + reflect_coeff * bwd_pred_error
-        bwd_pred_error = bwd_pred_error + reflect_coeff * fwd_pred_error_tmp
+        fwd_pred_error += reflect_coeff * bwd_pred_error
+        bwd_pred_error += reflect_coeff * fwd_pred_error_tmp
 
         # SNIP - we are now done with order M and advance. M-1 <- M
 
@@ -1101,10 +1101,10 @@ def zero_crossings(
         threshold = 0.0
 
     if callable(ref_magnitude):
-        threshold = threshold * ref_magnitude(np.abs(y))
+        threshold *= ref_magnitude(np.abs(y))
 
     elif ref_magnitude is not None:
-        threshold = threshold * ref_magnitude
+        threshold *= ref_magnitude
 
     if threshold > 0:
         y = y.copy()
@@ -1637,7 +1637,7 @@ def mu_expand(x, mu=255.0, quantize=True):
         )
 
     if quantize:
-        x = x * 2.0 / (1 + mu)
+        x *= 2.0 / (1 + mu)
 
     if np.any(x < -1) or np.any(x > 1):
         raise ParameterError(
diff --git a/librosa/core/constantq.py b/librosa/core/constantq.py
index 2291a62e1..594d20228 100644
--- a/librosa/core/constantq.py
+++ b/librosa/core/constantq.py
@@ -294,7 +294,7 @@ def hybrid_cqt(
         tuning = estimate_tuning(y=y, sr=sr, bins_per_octave=bins_per_octave)
 
     # Apply tuning correction
-    fmin = fmin * 2.0 ** (tuning / bins_per_octave)
+    fmin *= 2.0 ** (tuning / bins_per_octave)
 
     # Get all CQT frequencies
     freqs = cqt_frequencies(n_bins, fmin, bins_per_octave=bins_per_octave)
@@ -462,7 +462,7 @@ def pseudo_cqt(
         dtype = util.dtype_r2c(y.dtype)
 
     # Apply tuning correction
-    fmin = fmin * 2.0 ** (tuning / bins_per_octave)
+    fmin *= 2.0 ** (tuning / bins_per_octave)
 
     fft_basis, n_fft, _ = __cqt_filter_fft(
         sr,
@@ -624,7 +624,7 @@ def icqt(
         fmin = note_to_hz("C1")
 
     # Apply tuning correction
-    fmin = fmin * 2.0 ** (tuning / bins_per_octave)
+    fmin *= 2.0 ** (tuning / bins_per_octave)
 
     # Get the top octave of frequencies
     n_bins = C.shape[-2]
@@ -891,7 +891,7 @@ def vqt(
         dtype = util.dtype_r2c(y.dtype)
 
     # Apply tuning correction
-    fmin = fmin * 2.0 ** (tuning / bins_per_octave)
+    fmin *= 2.0 ** (tuning / bins_per_octave)
 
     # First thing, get the freqs of the top octave
     freqs = cqt_frequencies(n_bins, fmin, bins_per_octave=bins_per_octave)[
diff --git a/librosa/core/pitch.py b/librosa/core/pitch.py
index 0f645b226..3ba607d62 100644
--- a/librosa/core/pitch.py
+++ b/librosa/core/pitch.py
@@ -932,7 +932,7 @@ def __pyin_helper(
 
     # Refine peak by parabolic interpolation.
     period_candidates = min_period + yin_period
-    period_candidates = period_candidates + parabolic_shifts[yin_period, frame_index]
+    period_candidates += parabolic_shifts[yin_period, frame_index]
     f0_candidates = sr / period_candidates
 
     # Find pitch bin corresponding to each f0 candidate.
diff --git a/librosa/core/spectrum.py b/librosa/core/spectrum.py
index a4aa0f174..fe9e8704b 100644
--- a/librosa/core/spectrum.py
+++ b/librosa/core/spectrum.py
@@ -1281,7 +1281,7 @@ def phase_vocoder(D, rate, hop_length=None, n_fft=None):
         dphase = np.angle(columns[..., 1]) - np.angle(columns[..., 0]) - phi_advance
 
         # Wrap to -pi:pi range
-        dphase = dphase - 2.0 * np.pi * np.round(dphase / (2.0 * np.pi))
+        dphase -= 2.0 * np.pi * np.round(dphase / (2.0 * np.pi))
 
         # Accumulate phase
         phase_acc += phi_advance + dphase
diff --git a/librosa/display.py b/librosa/display.py
index 11f083fa7..b6dd69c3b 100644
--- a/librosa/display.py
+++ b/librosa/display.py
@@ -1336,6 +1336,6 @@ def __coord_cqt_hz(n, fmin=None, bins_per_octave=12, sr=22050, **_kwargs):
         fmin = core.note_to_hz("C1")
 
     # Apply tuning correction
-    fmin = fmin * 2.0 ** (_kwargs.get("tuning", 0.0) / bins_per_octave)
+    fmin *= 2.0 ** (_kwargs.get("tuning", 0.0) / bins_per_octave)
 
     # we drop by half a bin so that CQT bins are centered vertically
diff --git a/librosa/feature/inverse.py b/librosa/feature/inverse.py
index 8d004913b..844aeffe3 100644
--- a/librosa/feature/inverse.py
+++ b/librosa/feature/inverse.py
@@ -251,7 +251,7 @@ def mfcc_to_mel(mfcc, n_mels=128, dct_type=2, norm="ortho", ref=1.0, lifter=0):
             )
 
         # lifter mfcc values
-        mfcc = mfcc / (lifter_sine + tiny(mfcc))
+        mfcc /= lifter_sine + tiny(mfcc)
 
     elif lifter != 0:
         raise ParameterError("MFCC to mel lifter must be a non-negative number.")
diff --git a/librosa/feature/utils.py b/librosa/feature/utils.py
index 06ba056e9..cf3fbd2c0 100644
--- a/librosa/feature/utils.py
+++ b/librosa/feature/utils.py
@@ -248,7 +248,7 @@ def stack_memory(data, n_steps=2, delay=1, **kwargs):
 
     # Construct the shape of the target array
     shape = list(data.shape)
-    shape[-2] = shape[-2] * n_steps
+    shape[-2] *= n_steps
     shape[-1] = t
     shape = tuple(shape)
 
diff --git a/librosa/filters.py b/librosa/filters.py
index 59a845db4..70f6bbf46 100644
--- a/librosa/filters.py
+++ b/librosa/filters.py
@@ -560,7 +560,7 @@ def constant_q(
         )
 
         # Apply the windowing function
-        sig = sig * __float_window(window)(len(sig))
+        sig *= __float_window(window)(len(sig))
 
         # Normalize
         sig = util.normalize(sig, norm=norm)
diff --git a/librosa/onset.py b/librosa/onset.py
index 97d567926..87382daf0 100644
--- a/librosa/onset.py
+++ b/librosa/onset.py
@@ -151,7 +151,7 @@ def onset_detect(
     # (a common normalization step to make the threshold more consistent)
     if normalize:
         # Normalize onset strength function to [0, 1] range
-        onset_envelope = onset_envelope - onset_envelope.min()
+        onset_envelope -= onset_envelope.min()
         # Max-scale with safe division
         onset_envelope /= np.max(onset_envelope) + util.tiny(onset_envelope)
 
diff --git a/librosa/util/utils.py b/librosa/util/utils.py
index 5989beb63..a46df21bd 100644
--- a/librosa/util/utils.py
+++ b/librosa/util/utils.py
@@ -1247,7 +1247,7 @@ def peak_pick(x, pre_max, post_max, pre_avg, post_avg, delta, wait):
     detections = x * (x == mov_max)
 
     # Then mask out all entries less than the thresholded average
-    detections = detections * (detections >= (mov_avg + delta))
+    detections *= (detections >= (mov_avg + delta))
 
     # Initialize peaks array, to be filled greedily
     peaks = []
diff --git a/tests/test_core.py b/tests/test_core.py
index 612eb3556..b65078b98 100644
--- a/tests/test_core.py
+++ b/tests/test_core.py
@@ -550,7 +550,7 @@ def test_reassigned_spectrogram_flags(reassign_frequencies, reassign_times, cent
         frame_times = librosa.frames_to_time(np.arange(freqs.shape[1]))
 
         if not center:
-            frame_times = frame_times + (2048.0 / 22050.0 / 2.0)
+            frame_times += 2048.0 / 22050.0 / 2.0
 
         assert np.array_equiv(times, frame_times[np.newaxis, :])
 
@@ -1024,7 +1024,7 @@ def test_pyin_multi():
     h = librosa.filters.get_window('triangle', y.shape[-1])
 
     # Filter it
-    y = y * h[np.newaxis,:]
+    y *= h[np.newaxis,:]
 
     # Disable nans so we can use allclose checks
     fall, vall, vpall = librosa.pyin(y, fmin=100, fmax=1000, center=False, fill_na=-1)
@@ -1046,7 +1046,7 @@ def test_pyin_multi_center():
     h = librosa.filters.get_window('triangle', y.shape[-1])
 
     # Filter it
-    y = y * h[np.newaxis,:]
+    y *= h[np.newaxis,:]
 
     # Disable nans so we can use allclose checks
     fleft, vleft, vpleft = librosa.pyin(y, fmin=100, fmax=1000, center=False, fill_na=-1)
diff --git a/tests/test_filters.py b/tests/test_filters.py
index c8548ead5..0e2908ba5 100644
--- a/tests/test_filters.py
+++ b/tests/test_filters.py
@@ -278,7 +278,7 @@ def test_constant_q(sr, fmin, n_bins, bins_per_octave, filter_scale, pad_fft, no
     # Check for vanishing negative frequencies
     F_fft = np.abs(np.fft.fft(F, axis=1))
     # Normalize by row-wise peak
-    F_fft = F_fft / np.max(F_fft, axis=1, keepdims=True)
+    F_fft /= np.max(F_fft, axis=1, keepdims=True)
     assert not np.any(F_fft[:, -F_fft.shape[1] // 2 :] > 1e-4)
 
 
diff --git a/tests/test_met_features.py b/tests/test_met_features.py
index 3dd09b97f..b552b5c79 100644
--- a/tests/test_met_features.py
+++ b/tests/test_met_features.py
@@ -37,7 +37,7 @@ def met_stft(y, n_fft, hop_length, win_length, normalize):
     )
 
     if normalize:
-        S = S / (S[0] + np.sum(2 * S[1:], axis=0))
+        S /= S[0] + np.sum(2 * S[1:], axis=0)
 
     return S
 
diff --git a/tests/test_multichannel.py b/tests/test_multichannel.py
index 7f0852cfb..7d07a323a 100644
--- a/tests/test_multichannel.py
+++ b/tests/test_multichannel.py
@@ -1001,7 +1001,7 @@ def test_trim_multichannel(y_multi):
     y, sr = y_multi
 
     # Make one channel much quieter than the other
-    y = y * np.array([[1e-6, 1e6]]).T
+    y *= np.array([[1e-6, 1e6]]).T
     yt, ival = librosa.effects.trim(y)
 
     yt0, ival0 = librosa.effects.trim(y[0])
```


>…, I'll entertain specific suggestions with good justification, …

You added the information that existing software documentation (for a specific Python feature) would still be not good enough for you according to this clarification attempt, didn't you?
I am unsure under which circumstances development interests will grow a bit more.

**Comment by bmcfee** [comment]
> Interested developers can try the shown search and replace approach out, can't they?

Yes, but you could also have skipped this entire conversation by describing the proposed changes it would identify.

Asking devs to copypaste a baroque perl one-liner, clever as it might be, is pretty unreasonable.  In general, I would never ask anyone to run code in a shell that isn't 100% clear and obvious, if only for security reasons.  Anyone asking strangers to paste commands into a shell should be greeted with healthy suspicion, as I'm doing here.

> Would you find the following data format representation more appealing?

Yes!  This way we can at least discuss the merits of individual changes.  (Though, really, permalinking to the altered lines would be even better, but I'll take it.)

Now, going point by point:

- beat.py:normalize: definitely not, it would alter the input onset strength signal.

- beat.py:dp
    - candidates: this one would be fine, since it's a locally constructed variable
    - window: this would be okay as well
- audio.py:
    -  n: sure, but it wouldn't really make any difference (n is a scalar not an array)
    - n_samples: same
    - fwd_pred_error: nope, in this case we'd clobber the temporary storage and the result would be incorrect
    - bwd_pred_error: might be okay, but it's not obvious to me
    - threshold: should be fine, scalar
    - mu-law quantize: nope, side effects
- constantq:
    - fmin: okay
- pitch.py:
    - period_candidates: maybe okay.  there might be some subtleties to array broadcasting here
- spectrum.py:
    - dphase: probably okay, local variable
- display.py:
    - fmin: should be fine (scalar) but no obvious benefit
- filters.py:
    - constantq: probably okay, it's a local variable being constructed here
- onset.py:
    - onset_envelope: nope, side effects
- utils.py:
    - detections: okay, local variable
- almost everything in tests: nope, modifying fixtures in-place would invalidate different runs of the tests

So of all that diff, there are maybe 6 lines that could be changed as you suggest.  The rest would render the code incorrect (eg by modifying input arrays, or unsafely modifying other objects that are multiply referenced locally) or provide no clear benefit (ie relate to scalars instead of arrays).

**Comment by elfring** [comment]
>beat.py:normalize: definitely not, it would alter the input onset strength signal.

[The variable “onsets”](https://github.com/librosa/librosa/blob/cd9e1568a74b8ee84e2750fc2f1aedc2ef5d7ce8/librosa/beat.py#L637 "Update candiate: __normalize_onsets()") gets reassigned (as usual).


>fwd_pred_error: nope, in this case we'd clobber the temporary storage and the result would be incorrect

I imagine that different development views can need another bit of clarification at [such a source code place](https://github.com/librosa/librosa/blob/eb603e7a91598d1e72d3cdeada0ade21a33f9c0c/librosa/core/audio.py#L974 "Update candiate: __lpc()").


>almost everything in tests: nope, modifying fixtures in-place would invalidate different runs of the tests

* May this view be reconsidered?
* How do you think about to add a few test cases for further comparisons and checks of the discussed software behaviour?


>The rest would render the code incorrect (eg by …)

Would any other developers or code reviewers get into the mood to check such information in more detail?


>or provide no clear benefit (ie relate to scalars instead of arrays).

* Can this implementation detail trigger further software insights?
* Would you occasionally care for [the software design principle “Don't repeat yourself”](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself "Explanation") (also according to these computations)?

**Comment by bmcfee** [comment]
> [The variable “onsets”](https://github.com/librosa/librosa/blob/cd9e1568a74b8ee84e2750fc2f1aedc2ef5d7ce8/librosa/beat.py#L637) gets reassigned (as usual).

Yes, that's the point.  If it was changed to an in-place update (`/=`), since arrays are passed by reference, it would modify the input.



> I imagine that different development views can need another bit of clarification at [such a source code place](https://github.com/librosa/librosa/blob/eb603e7a91598d1e72d3cdeada0ade21a33f9c0c/librosa/core/audio.py#L974).

I don't know what you mean by this.  It's not a matter of opinion here: changing to an in-place update would be incorrect, and obviously so.



> May this view be reconsidered?

No.  Tests should never modify fixtures.


> How do you think about to add a few test cases for further comparisons and checks of the discussed software behaviour?

That seems overkill to me.  Our tests are already expensive enough, and adding another layer of testing here would significantly increase the computational expense, and be hard to implement in general.  I would much rather be rigorous about enforcing a lack of side-effects at code review time.



> Would any other developers or code reviewers get into the mood to check such information in more detail?

I doubt it.  This is a small project, and we're all pretty busy.



> * Can this implementation detail trigger further software insights?
> 

I don't think so?  This is all well-trodden territory, and everything here is pretty clear if you take the time to understand:

1. How python function calls pass data around (by reference)
2. The interface of numpy ndarrays (at a high level at least)
3. The typical conventions adopted by the scientific python community

> * Would you occasionally care for [the software design principle “Don't repeat yourself”](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) (also according to these computations)?

I don't think DRY applies in any of the cases identified.  We're not talking about redundant code here, we're talking about the difference between updating an object in-place and deriving a new object by applying a function (eg addition).  At the risk of repeating myself (again) here: sometimes an in-place update is appropriate, and sometimes it's not, and you really have to understand the context of the code in question to make that determination.

At this point, I think we've discussed this issue enough, and I don't see much point to pursuing it further.  Yes there are some places that could be improved.  If you'd like to take that on, please submit a pull request.  If not, I'll close this issue out.

**Comment by elfring** [comment]
>If it was changed to an in-place update (`/=`), since arrays are passed by reference, it would modify the input.

I guess that this view about the involved data flow will need further clarification.


>… changing to an in-place update would be incorrect, and obviously so.

I got an other impression about such functionality from the Python programming language.


>Our tests are already expensive enough, and adding another layer of testing here would significantly increase the computational expense,

This would be usual, wouldn't it?


>and be hard to implement in general.

May any test cases demonstrate the applicability of in-place modifications (also for higher level data structures like mentioned NumPy arrays)? :thinking:


>We're not talking about redundant code here, …

:thought_balloon: Would augmented assignment statements have got the effect that repeated specifications would be reduced a bit for selected variable names?


>Yes there are some places that could be improved.

Thanks for this acknowledgement.

**Comment by bmcfee** [comment]
> I got an other impression about such functionality from the Python programming language.

If that's the case, then I would recommend becoming more familiar with the language and libraries (i.e. numpy) before proposing changes to existing code-bases.  Better yet, make an effort to read and understand a project's code before proposing modifications!  You'll learn a bunch and save everyone involved quite a bit of time and frustration.

Don't be surprised if maintainers find it rude and annoying when a would-be contributor barges into the issues proposing changes that have not been thought out.

> This would be usual, wouldn't it?

I don't understand your point here.  I'm all for testing - moreso than most! - but only if it's for a specific goal.  I don't see much point in adding mutability tests everywhere if we can enforce correct behavior by other means.


> May any test cases demonstrate the applicability of in-place modifications

I don't understand your question.  From my perspective, I can't think of any situations in which library code should be mutating input data in-place unless the user has specifically requested that (eg by using a `out=` parameter in numpy functions).


> thought_balloon Would augmented assignment statements have got the effect that repeated specifications would be reduced a bit for selected variable names?

No?  Take a look at the specific instances where this might be applicable.  You maybe save a few characters of code, but that's quite distant from the motivation of DRY.  More often than not, the more verbose code is easier to read, and all else being equal, is preferable.

**Comment by elfring** [comment]
>From my perspective, I can't think of any situations in which library code should be mutating input data in-place …

:thought_balloon: I imagine that design goals like reduced memory allocations can occasionally become also desirable.

Both assignment variants can support correct data processing.
Test cases can eventually help to measure different software run time characteristics accordingly.
Concrete measurements will probably point interesting system settings out, won't they? :thinking:

**Comment by bmcfee** [comment]
> Both assignment variants can support correct data processing.

Nobody is disputing that.  As I've said **multiple times**, this needs to be determined on a case by case basis because more often than not, they actually do not behave identically.  I don't know what about this is unclear to you, or why you insist on continuing this conversation instead of just making the PR if you're so invested in the idea.


I don't think this is a productive conversation anymore, so I'm closing this issue.
